clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -779,14 +779,25 @@ public synchronized void maybeTransitionToErrorState(RuntimeException exception)
     }
 
     synchronized void handleFailedBatch(ProducerBatch batch, RuntimeException exception, boolean adjustSequenceNumbers) {
-        maybeTransitionToErrorState(exception);
+        // Compare the batch with the current ProducerIdAndEpoch. If the producer IDs are the *same* but the epochs
+        // are *different*, consider the batch as stale.
+        boolean isStaleBatch = batch.producerId() == producerIdAndEpoch.producerId && batch.producerEpoch() != producerIdAndEpoch.epoch;
```

Comment:
one minor thing here, if we have an epoch overflow, we won't consider this check. Maybe if the producer ID doesn't match, we also ignore it? Not sure if that has unintended consequences. We could also tackle as a followup, since without this fix, it's the same state as today.

Reply:
I think I'm missing something conceptually—when an epoch overflows, does it restart at 0, roll over to negative, or something else?

In the case of an overflow, wouldn't the batch's producer epoch value still differ from the `producerIdAndEpoch`’s value?

Reply:
When we have epoch overflow, we get a new producer id and start epoch at 0.

Reply:
So the producer ID will not be the same.

Reply:
So if _either_ the producer ID _or_ the epoch of the batch is different than that of the transaction manager, it's "stale?" 

```suggestion
        boolean isStaleBatch = batch.producerId() != producerIdAndEpoch.producerId || batch.producerEpoch() != producerIdAndEpoch.epoch;
```

Reply:
That could work -- I guess the other thing is do we check and handle if the producer ID doesn't match elsewhere? I don't remember off the top of my head.

Reply:
Here are the places I found in which a `ProducerIdAndEpoch` is compared:

* [`TransactionManager.setProducerIdAndEpoch()`](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L604-L615) checks the `producerId`, but it appears to only affect logging, though.
* [`TransactionManager.maybeUpdateProducerIdAndEpoch()`](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L589-L602) calls to `hasStaleProducerIdAndEpoch()` to compare its current `ProducerIdAndEpoch` with the one in its `txnPartitionMap`. In the case we're seeing, the producer ID in the `ProducerBatch` is out of sync. I don't know if the `txnPartitionMap` is also out of sync in that case.
* [`ProducerBatch.resetProducerState()`](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java#L476-L481) seems interesting to consider in that maybe it could be called out of sync with the transaction manager? That method is called by [`TxnPartitionEntry.adjustSequencesDueToFailedBatch()`](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/TxnPartitionEntry.java#L139-L152), but resets the batch with the same `ProducerIdAndEpoch` from the batch. Should it be consulting the `TransactionManager` for the _current_ value?

Reply:
Not sure I follow the last point. That method is always called from the TransactionManager's `txnPartitionMap`

Reply:
Anyway, this is not quite what I meant. I meant, do we check in the failedBatch path anywhere. Doing a quick search now, it appears not.

Reply:
I think that means it is safe to mark as stale if the producer ID doesn't match

Reply:
I've made the change to check the producer ID too. PTAL. Thanks!

Reply:
@jolshan—There are a number of tests that fail with the latest change:

* `TransactionManagerTest.testTransitionToAbortableErrorOnBatchExpiry()`
* `TransactionManagerTest.testTransitionToAbortableErrorOnMultipleBatchExpiry()`
* `TransactionManagerTest.testBackgroundInvalidStateTransitionIsFatal()`
* `TransactionsTest.testBumpTransactionalEpochWithTV2Disabled()`
* `TransactionsWithTieredStoreTest.testBumpTransactionalEpochWithTV2Disabled()`

From the perspective of those tests, the state should be set to `FATAL_ERROR` in the case where the batch’s producer ID is different from the transaction manager’s.

Reply:
Ok. So I suppose this was being check somewhere else 😅 I suppose we can revert that. 
Perhaps it is sufficient to still fail in the edge case where we have epoch overflow for now as it doesn't seem like we will make much progress on finding an alternative.

Reply:
Is `batch.producerEpoch() != producerIdAndEpoch.epoch` ok? Or should it be `batch.producerEpoch() < producerIdAndEpoch.epoch`?

On the other hand, we don't have a check here for batch.producerEpoch() > producerIdAndEpoch.epoch` which should never happen? Should we add such a check (or does it maybe exist somewhere else)? Or do we not care?

Reply:
Receiving a future epoch would seem to imply something was really wrong.

Reply:
So it sounds like `batch.producerEpoch() < producerIdAndEpoch.epoch` is really more appropriate as a "staleness"  check and let the `TransactionManager` handle the other cases?

Reply:
I think it is possible that these stale overflow cases unfortunately may lead to having to close the producer, but I would prefer to look into that as a potential followup. Seems like it would be very rare.